File: pa5/Dictionary.c
#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<assert.h>
#include"Dictionary.h"


// private types --------------------------------------------------------------

//arbitrary value
const int tableSize = 101;

// rotate_left()
// rotate the bits in an unsigned int
unsigned int rotate_left(unsigned int value, int shift)
{
    int sizeInBits = 8 * sizeof(unsigned int);
    shift = shift & (sizeInBits - 1);
    if ( shift == 0 )
        return value;
    return (value << shift) | (value >> (sizeInBits - shift));
}

// pre_hash()
// turn a string into an unsigned int
unsigned int pre_hash(char *input)
{
    unsigned int result = 0xBAE86554;
    while (*input)
    {
        result ^= *input++;
        result = rotate_left(result, 5);
    }
    return result;
}

// hash()
// turns a string into an int in the range 0 to tableSize-1
int hash(char *key)
{
    return pre_hash(key) % tableSize;
}
// NodeObj
typedef struct NodeObj
{
    char *key;
    char *value;
    struct NodeObj *next;
} NodeObj;

// Node
typedef NodeObj *Node;

// newNode()
// constructor of the Node type
Node newNode(char *k, char *v)
{
    Node N = malloc(sizeof(NodeObj));
    assert(N != NULL);
    N->key = k;
    N->value = v;
    N->next = NULL;
    return (N);
}

// freeNode()
// destructor for the Node type
void freeNode(Node *pN)
{
    if ( pN != NULL && *pN != NULL )
    {
        free(*pN);
        *pN = NULL;
    }
}

// LinkedObj              //used to keep track of the list
typedef struct LinkedObj
{
    Node head;
} LinkedObj;

// Linked List
typedef LinkedObj *LinkedList;

// Constructor for the LinkedList type
LinkedList newLinkedList()
{
    LinkedList L = malloc(sizeof(LinkedObj));
    L->head = NULL;
    return L;
}

typedef struct DictionaryObj
{
    LinkedList table;
    int numPairs;
} DictionaryObj;

//finds key within the hash table
Node findKey(Dictionary D, char *k)
{
    //traverse through nodes to check if node contains the key
    for (Node N = D->table[hash(k)].head; N != NULL; N = N->next) //run through the hash to find the key

        if (strcmp(k, N->key) == 0)    //check if the key looked for is found and is if found return the key and its value
        {
            return N;
        }
    return NULL; //key hasn't been found so just return null
}


// public functions -----------------------------------------------------------

// newDictionary()
// constructor for the Dictionary type
Dictionary newDictionary(void)
{
    Dictionary D = malloc(sizeof(DictionaryObj));
    assert(D != NULL);
    D->table = calloc(tableSize, sizeof(LinkedObj));    //allocate the array size for the linked lists
    D->numPairs = 0;
    return D;
}

// freeDictionary()
// destructor for the Dictionary type
void freeDictionary(Dictionary *pD)
{
    if ( pD != NULL && *pD != NULL )
    {
        if ( !isEmpty(*pD) ) makeEmpty(*pD);
        free(*pD);
        *pD = NULL;
    }
}

// isEmpty()
// returns 1 (true) if S is empty, 0 (false) otherwise
// pre: none
int isEmpty(Dictionary D)
{
    if (D->numPairs == 0)
    {
        return (1);
    }
    return 0;
}

int size(Dictionary D)
{
    //check if pointing to NULL
    if (D == NULL)
    {
        fprintf(stderr, "Dictionary Error: Calling size() on null pointer.\n");
    }
    return (D->numPairs);

}

// lookup()
// returns the value v such that (k, v) is in D, or returns NULL if no
// such value v exists.
// pre: none
char *lookup(Dictionary D, char *k)
{
    if ( D == NULL )
    {
        fprintf(stderr, "Dictionary Error: calling lookup() on NULL Dictionary reference.\n");
        exit(EXIT_FAILURE);
    }

    if ( D->numPairs == 0 )
    {
        fprintf(stderr, "Dictionary Error: calling lookup() on empty Dictionary list.\n");
        exit(EXIT_FAILURE);
    }
    if (findKey(D, k) != NULL)    //searches for the value using hashing
    {
        return findKey(D, k)->value;
    }
    else
    {
        return NULL;
    }
}

// insert()
// inserts new (key,value) pair into D
// pre: lookup(D, k)==NULL
void insert(Dictionary D, char *k, char *v)
{
    if ( D == NULL )
    {
        fprintf(stderr, "Dictionary Error: calling insert() on NULL Dictionary reference\n");
        exit(EXIT_FAILURE);
    }
    if (findKey(D, k) != NULL)
    {
        fprintf(stderr, "Dictionary Error: calling insert() on a duplicate key: %s\n", k );
        exit(EXIT_FAILURE);
    }

    //passes both the error checks so now we insert the key into the array of linked lists
    if (D->table[hash(k)].head == NULL)     //if the array is empty, then we insert a new node for a linkedList
    {
        Node N = newNode(k, v);
        N->next = D->table[hash(k)].head;   //travels down through the table to insert keys and values
        D->table[hash(k)].head = N;         //set new head
        D->numPairs++;                      //increment the numPairs
    }
    else
    {
        if (lookup(D, k) == NULL)             //make sure value isn't already there
        {
            //if the index in where we want to store our keys and value in
            //are taken, we keep traversing through the linked address to find a null
            //pointer and then chain the key and value in there
            for (Node N = D->table[hash(k)].head; N->next != NULL; N = N->next)
            {
                N->next = newNode(k, v);
            }
            D->numPairs++;      //increment the numPiars
        }
    }
}

// delete()
// deletes pair with the key k
// pre: lookup(D, k)!=NULL
void delete(Dictionary D, char *k)
{
    if (D == NULL)
    {
        fprintf(stderr, "Dictionary Error: calling delete() on a NULL Dictionary reference.\n");
        exit(EXIT_FAILURE);
    }
    if (lookup(D, k) == NULL)
    {
        fprintf(stderr, "Dictionary Error: calling delete() on a key not even in the table.\n");
        exit(EXIT_FAILURE);
    }
    //now that we passed the error checks, we go about deleting nodes from the table of linked lists
    Node N = D->table[hash(k)].head;
    if (N == findKey(D, k))       //found the key but check for its next elements in the list
    {
        if (D->table[hash(k)].head->next == NULL)    //there is nothing ahead of this list so
        {
            freeNode(&D->table[hash(k)].head);    //we can just free it without loss of memory
            D->numPairs--;                  //decrement the numPairs
        }
        //there are some elements ahead of the found key so we need to move the head without losing the list
        else
        {
            N = D->table[hash(k)].head;
            D->table[hash(k)].head = D->table[hash(k)].head->next; //set its new head so we can get rid of the element before us
            freeNode(&N);
            D->numPairs--;
        }
    }
}


// makeEmpty()
// re-sets D to the empty state.
// pre: none
void makeEmpty(Dictionary D)
{
    for (int i = 0; i < tableSize; i++)
    {
        Node N = D->table[i].head;
        Node T; //create a temporary node to save the position
        while (N != NULL)
        {
            T = N;
            N = N->next;
            freeNode(&T);
            D->numPairs--;
        }
    }
    free(D->table);
}


// printDictionary()
// pre: none
// prints a text representation of D to the file pointed to by out
void printDictionary(FILE *out, Dictionary D)
{
    if (D == NULL)
    {
        fprintf(stderr, "Dictionary Error: calling printDictionary() on NULL Dictionary reference\n");
        exit(EXIT_FAILURE);
    }
    for (int i = 0 ; i < tableSize; i++)
    {
        for (Node N = D->table[i].head; N != NULL; N = N->next)
        {
            fprintf(out, "%s %s\n", N->key, N->value);    //print out the values
        }
    }
}

File: labs/lab7/Dictionary.java
//Steven Cruz
//cs 12b
//lab 7
//sicruz

public class Dictionary implements DictionaryInterface {


    //private inner Node class
    private class Node {
        String key;
        String value;
        Node left;
        Node right;

        //constructor for the node
        Node(String k, String v) {
            key = k;
            value = v;
            left = null;
            right = null;

        }
    }
    //Fields for the Dictionary class
    private Node left;     // reference to left node of the list
    private Node right;    // reference to the right node of the list
    private Node root;
    private int numPairs; //number of keys in the Dictionary

    // findKey()
    // returns the Node containing key k in the subtree rooted at R, or returns
    // NULL if no such Node exists
    private Node findKey(Node R, String key) {
        if (R == null || key.compareTo(R.key) == 0) {
            return R;
        }
        if (key.compareTo(R.key) < 0) {
            return findKey(R.left, key);
        } else {
            return findKey(R.right, key);
        }
    }

    // findParent()
    // returns the parent of N in the subtree rooted at R, or returns NULL
    // if N is equal to R. (pre: R != NULL)
    private Node findParent(Node N, Node R) {
        Node P = null;
        if (N != R) {
            P = R;
            while (P.left != N && P.right != N) {
                if (N.key.compareTo(P.key) < 0)
                    P = P.left;
                else
                    P = P.right;
            }
        }
        return P;
    }

    // findLeftmost()
    // returns the leftmost Node in the subtree rooted at R, or NULL if R is NULL.
    private Node findLeftmost(Node R) {
        Node L = R;
        if (L != null) for (; L.left != null; L = L.left);
        return L;
    }

    // printInOrder()
    // prints the (key, value) pairs belonging to the subtree rooted at R in order
    // of increasing keys to file pointed to by out.
    private String printInOrder(Node R) {
        String a, b, c;
        a = b = c = ""; 
        if (R != null) {
            a = printInOrder(R.left);
            b = R.key + " " + R.value + "\n"; 
            c = printInOrder(R.right);
        }
        return (a+b+c); 
    }

    // deleteAll()
    // deletes all Nodes in the subtree rooted at N.
    private void deleteAll(Node N) {
        if (N != null) {
            deleteAll(N.left);
            deleteAll(N.right);
        }
    }


    //public functions

    //Dictionary()
    //constructor for the Dictionary Class
    public Dictionary() {
        Node root = null;
        numPairs = 0;
    }

    // isEmpty
    // pre: none
    // post: returns true if Dictionary contains no pairs. Returns false otherwise.
    public boolean isEmpty() {
        return (numPairs == 0);
    }

    //size()
    //pre: none
    //post: returns the number of pairs in the Dictionary
    public int size() {
        if (numPairs == 0) {
            return 0;
        } else {
            return numPairs;
        }
    }

    //lookup
    //pre: none
    //post: returns value associated key, or null reference if no such key exists
    public String lookup(String key) {
        Node N = findKey(root, key);
        if (N != null) {
            return N.value;
        } else {
            return null;
        }
    }

    // insert()
    // inserts new (key,value) pair into this Dictionary
    // pre: lookup(key)==null
    // post: inserts new (key,value)
    public void insert(String key, String value) throws KeyCollisionException {
        Node N, A, B;
        if (findKey(root, key) != null) {
            throw new KeyCollisionException (
                "insert(): cannot insert duplicate keys.");
        }

        N = new Node(key, value);
        B = null;
        A = root;
        while (A != null) {
            B = A;
            if (key.compareTo(A.key) < 0) A = A.left;
            else A = A.right;
        }
        if (B == null) root = N;
        else if (key.compareTo(B.key) < 0) B.left = N;
        else B.right = N;
        numPairs++;
    }

    // delete()
    // deletes pair with the given key
    // pre: lookup(key)!=null
    // post: removes a pair with the given key
    public void delete(String key) throws KeyNotFoundException {
        Node N, P, S;
        N = findKey(root, key);
        if (findKey(root, key) == null) {
            throw new KeyNotFoundException (
                "cannot delete non-existent key.");
        }
        if (N.left == null && N.right == null) {    //case 1: N has no children
            if (N == root) {
                root = null;
            } else {
                P = findParent(N, root);
                if (P.right == N) P.right = null;
                else P.left = null;
            }
        }
        else if (N.right == null) {                      //case 2.1: N has a left child
            if (N == root) {
                root = N.left;
            } else {
                P = findParent(N, root);
                if (P.right == N) P.right = N.left;
                else P.left = N.left;
            }
        }
        else if (N.left == null) {                      //case 2.2 N has a right child
            if (N == root) {
                root = N.right;
            } else {
                P = findParent(N, root);
                if (P.right == N) P.right = N.right;
                else P.left = N.right;
            }
        }



        else { //N.left != null && N.right != null   case 3 where N has both left and right childs
            S = findLeftmost(N.right);
            N.key = S.key;
            N.value = S.value;
            P = findParent(S, N);
            if (P.right == S) P.right = S.right;
            else P.left = S.right;

        }
        numPairs--;
    }

    // makeEmpty()
    // pre: none
    // post: isEmpty()
    public void makeEmpty() {
        deleteAll(root);
        root = null;
        numPairs = 0;
    }


   // toString()
   // returns a String representation of this Dictionary
   // overrides Object's toString() method
   // pre: none
    public String toString() {
        String s = "";
        s = printInOrder(root);
        return s;
    }
}

